daemon/containerd/image_pull.go

```diff
@@ -68,6 +70,29 @@ func (i *ImageService) PullImage(ctx context.Context, image, tagOrDigest string,
 	infoHandler := snapshotters.AppendInfoHandlerWrapper(ref.String())
 	opts = append(opts, containerd.WithImageHandlerWrapper(infoHandler))
 
-	_, err = i.client.Pull(ctx, ref.String(), opts...)
+	img, err := i.client.Pull(ctx, ref.String(), opts...)
+	if err != nil {
+		return err
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
+		"imageID": img.Target().Digest,
+		"tag":     ref.String(),
+	})
+	logger.Info("image pulled")
+
+	// The pull succeeded, check if we have a dangling image for this target
+	danglingImage, err := i.client.ImageService().Get(context.Background(), danglingImageName(img.Target().Digest))
+	if err != nil {
+		if !cerrdefs.IsNotFound(err) {
+			logger.WithError(err).Warn("unexpected error when looking for dangling image")
+		}
+		return nil
+	}
+
+	// Delete the dangling image, as it's no longer dangling.
+	if err := i.client.ImageService().Delete(context.Background(), danglingImage.Name); err != nil {
+		logger.WithError(err).Warn("unexpected error when deleting dangling image")
+	}
 	return err
```

Comment:
This `return err` will always be `nil`; the `Delete()` block is shadowing the former error produced by `Get()`.


```suggestion
	err = i.client.ImageService().Delete(context.Background(), danglingImage.Name)
	if err != nil && !cerrdefs.IsNotFound(err) {
		// The pull succeeded, but cleaning up the dangling image failed. Ignore the
		// error to not mark the pull as failed.
		logger.WithError(err).Warn("unexpected error when deleting dangling image")
	}
	return nil
```

or (if we remove the earlier check); probably should be using `danglingImageName(img.Target().Digest)` (or a variant thereof?)


❓ Are there any errors we _should_ actually return in this case?

Reply:
I don't think so – at this point the pull succeeded and using the image/subsequent commands relying on it will work fine, so erroring would be a bit weird, we just failed at some "cleanup/side" tasks related to the pull.

Reply:
Yeah, I guess it's nothing actionable as well for the user. Mostly curious "what" could go wrong here. If such errors are really exceptional, it could make sense to surface them to the user (assuming "this never happens").

Maybe we can stream it back to the client as a warning 🤔 if we don't want things to fail (but still make them visible) 🤔 

☝️ that's definitely not for this PR though, but something we could explore as an option.